csharp/src/Ice/InputStream.cs

```diff
@@ -2845,33 +2717,118 @@ internal override void skipSlice()
                 b.position(end);
 
                 var info = new SliceInfo(typeId, compactId, bytes, Array.Empty<Value>(), hasOptionalMembers, isLastSlice);
+                _current.slices.Add(info);
 
-                if (_current.slices == null)
+                // The deferred indirection table is only used by classes. For exceptions, the indirection table is
+                // unmarshaled immediately.
+                if (_current.sliceType == SliceType.ValueSlice)
                 {
-                    _current.slices = new List<SliceInfo>();
-                    _current.indirectionTables = new List<int[]>();
+                    _current.DeferredIndirectionTableList ??= new List<int>();
+                    if ((_current.sliceFlags & Protocol.FLAG_HAS_INDIRECTION_TABLE) != 0)
+                    {
+                        _current.DeferredIndirectionTableList.Add(_stream.pos());
+                        SkipIndirectionTable();
+                    }
+                    else
+                    {
+                        _current.DeferredIndirectionTableList.Add(0);
+                    }
                 }
-
-                //
-                // Read the indirect instance table. We read the instances or their
-                // IDs if the instance is a reference to an already unmarshaled
-                // instance.
-                //
-                if ((_current.sliceFlags & Protocol.FLAG_HAS_INDIRECTION_TABLE) != 0)
+                else
                 {
-                    int[] indirectionTable = new int[_stream.ReadAndCheckSeqSize(1)];
-                    for (int i = 0; i < indirectionTable.Length; ++i)
+                    _current.IndirectionTableList ??= new List<int[]?>();
+                    if (_current.IndirectionTable != null)
```

Comment:
We could also check `_current.sliceFlags & Protocol.FLAG_HAS_INDIRECTION_TABLE) != 0` and assert on `_current.IndirectionTable != null` to make it clear that the table has been read by `startSlice`.

Reply:
Fixed as suggested.